core/main/src/broker/websocket_broker.rs

```diff
@@ -0,0 +1,110 @@
+    let url_split: Vec<&str> = url.split("://").collect();
+    if let Some(domain) = url_split.get(1) {
+        let domain_split: Vec<&str> = domain.split('/').collect();
+        domain_split.first().unwrap().to_string()
+    } else {
+        url.to_owned()
+    }
+}
+
+impl WebsocketBroker {
+    fn start(endpoint: RuleEndpoint, callback: BrokerCallback) -> Self {
+        let (tx, mut tr) = mpsc::channel(10);
+        let broker = BrokerSender { sender: tx };
+        tokio::spawn(async move {
+            info!("Broker Endpoint url {}", endpoint.url);
+            let url = url::Url::parse(&endpoint.url).unwrap();
+            let port = extract_tcp_port(&endpoint.url);
+            info!("Url host str {}", url.host_str().unwrap());
+            //let tcp_url = url.host_str()
+            let tcp = loop {
+                if let Ok(v) = TcpStream::connect(&port).await {
+                    break v;
+                } else {
+                    error!("Broker Wait for a sec and retry {}", port);
+                    tokio::time::sleep(Duration::from_secs(1)).await;
+                }
+            };
+
+            let (stream, _) = client_async(url, tcp).await.unwrap();
+            let (mut ws_tx, mut ws_rx) = stream.split();
+
+            tokio::pin! {
+                let read = ws_rx.next();
+            }
+            loop {
+                tokio::select! {
+                    Some(value) = &mut read => {
+                        match value {
+                            Ok(v) => {
+                                if let tokio_tungstenite::tungstenite::Message::Text(t) = v {
+                                    // send the incoming text without context back to the sender
+                                    Self::handle_response(t.as_bytes(),callback.clone())
+                                }
+                            },
+                            Err(e) => {
+                                error!("Broker Websocket error on read {:?}", e);
+                                break false
+                            }
+                        }
+
+                    },
+                    Some(request) = tr.recv() => {
+                        debug!("Got request from receiver for broker {:?}", request);
+                        if let Ok(updated_request) = Self::update_request(&request) {
+                            debug!("Sending request to broker {}", updated_request);
+                             let _feed = ws_tx.feed(tokio_tungstenite::tungstenite::Message::Text(updated_request)).await;
+                            let _flush = ws_tx.flush().await;
+                        }
+
+                    }
+                }
+            }
+        });
+        Self { sender: broker }
+    }
+}
+
+impl EndpointBroker for WebsocketBroker {
+    fn get_broker(endpoint: RuleEndpoint, callback: BrokerCallback) -> Self {
+        Self::start(endpoint, callback)
+    }
+
+    fn get_sender(&self) -> BrokerSender {
```

Comment:
All implementers do the same thing, maybe just add the method to the trait.

Reply:
Yes each implementer gets a unique broker sender which is used by endpoint broker state to send the request. 

BrokerSender is self sized to each instance of the Broker type. Super trait cannot hold instance of BrokerSender without having method definition for the same and it also needs a Self sized instance. This implementation is similar to the ExtnProcessor trait.